### PR TITLE
Fixing hard IP related schema limitations

### PR DESCRIPTION
**Pull request by aolofsson**

### CODE REVIEW

File: siliconcompiler/core.py [hunk 1]
```diff
@@ -1530,6 +1530,31 @@ def _check_files(self):
...
+        allkeys = self.getkeys()
+        for keypath in allkeys:
+            allpaths = []
+            paramtype = self.get(*keypath, field='type')
+            if 'file' in paramtype or 'dir' in paramtype:
+                if 'dir' not in keypath and self.get(*keypath):
+                    allpaths = list(self.get(*keypath))
+                for path in allpaths:
+                    #check for env var
+                    m = re.match(r'\$(\w+)(.*)', path)
```
Comment (WRansohoff) [inline siliconcompiler/core.py:1548]: Could we use the `_resolve_env_vars` method for this, or is the logic a bit different?

**Review by WRansohoff — APPROVED** [review]
Looks good to me, and I don't see any conflicts with the klayout/environment changes that I submitted earlier.

I wonder if ZeroSoc's floorplanning/build scripts might need updating to work with the new 'stackup' schema entries, but if it does I'm sure the changes would be minor.

File: siliconcompiler/core.py (hunk 1 above)
Comment (aolofsson) [reply siliconcompiler/core.py:1548]: Possibly...

File: siliconcompiler/schema.py [hunk 2]
```diff
@@ -580,8 +581,8 @@ def schema_pdk(cfg, stackup='default'):
             "api: chip.set('pdk','file','xyce','M10','spice','asap7.sp')"],
         'help': """
         List of named files specified on a per tool and per stackup basis.
-        The parameter is useful for specifying files that are not directly
-        covered by the rest of the PDK schema.
+        The parameter should only be used for specifying files that are
+        not directly  supported by the SiliconCompiler PDK schema.
```
Comment (nmoroze) [inline siliconcompiler/schema.py:585]:
```suggestion
        not directly supported by the SiliconCompiler PDK schema.
```

File: siliconcompiler/schema.py [hunk 3]
```diff
@@ -602,8 +603,8 @@ def schema_pdk(cfg, stackup='default'):
             "api: chip.set('pdk','directory','xyce','M10','rfmodel','rftechdir')"],
         'help': """
         List of named directories specified on a per tool and per stackup basis.
-        The parameter is useful for specifying directories that are not directly
-        covered by the rest of the PDK schema.
+        The parameter should only be used for specifying files that are
+        not directly  supported by the SiliconCompiler PDK schema.
```
Comment (nmoroze) [inline siliconcompiler/schema.py:607]:
```suggestion
        not directly supported by the SiliconCompiler PDK schema.
```

File: siliconcompiler/schema.py [hunk 4]
```diff
@@ -624,8 +625,8 @@ def schema_pdk(cfg, stackup='default'):
             "api: chip.set('pdk','variable','xyce', 'M10','modeltype','bsim4')"],
         'help': """
         List of key/value strings specified on a per tool and per stackup basis.
-        The parameter is useful for specifying values that are not directly
-        covered by the rest of the PDK schema.
+        The parameter should only be used for specifying variables that are
+        not directly  supported by the SiliconCompiler PDK schema.
```
Comment (nmoroze) [inline siliconcompiler/schema.py:629]:
```suggestion
        not directly supported by the SiliconCompiler PDK schema.
```

File: siliconcompiler/tools/openroad/sc_apr.tcl [hunk 5]
```diff
@@ -118,15 +120,15 @@ read_lef  $sc_techlef
...
+	read_lef [dict get $sc_cfg library $lib lef $sc_stackup]
 }
 
 # Read Macrolibs
 foreach lib $sc_macrolibs {
     if {[dict exists $sc_cfg library $lib nldm]} {
         read_liberty [dict get $sc_cfg library $lib nldm typical lib]
     }
-    read_lef [dict get $sc_cfg library $lib lef]
+    read_lef [dict get $sc_cfg library $lib lef stackup]
```
Comment (nmoroze) [inline siliconcompiler/tools/openroad/sc_apr.tcl:131]: Should this be `$sc_stackup`?

**Review by nmoroze — APPROVED** [review]
A few small typos, otherwise LGTM!

**Comment by nmoroze** [comment]
> I wonder if ZeroSoc's floorplanning/build scripts might need updating to work with the new 'stackup' schema entries, but if it does I'm sure the changes would be minor.

Yup, good point -- I'll take care of this.

File: siliconcompiler/tools/openroad/sc_apr.tcl (hunk 5 above)
Comment (aolofsson) [reply siliconcompiler/tools/openroad/sc_apr.tcl:131]: Makes sense, I will fix in the next round.

**Pull request merged**